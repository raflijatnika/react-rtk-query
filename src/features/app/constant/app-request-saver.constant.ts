export enum REQUEST_SAVER {
  // TODO
  TODO_FETCH_LIST_REQUEST = 'TODO_FETCH_LIST_REQUEST',
  TODO_FETCH_DETAIL_REQUEST = 'TODO_FETCH_DETAIL_REQUEST',
  TODO_CREATE_REQUEST = 'TODO_CREATE_REQUEST',
  TODO_UPDATE_REQUEST = 'TODO_UPDATE_REQUEST',
  TODO_DELETE_REQUEST = 'TODO_DELETE_REQUEST'
}
