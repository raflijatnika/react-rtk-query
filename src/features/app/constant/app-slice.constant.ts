// Interfaces
import { IAppSliceInitialState } from '@/features/app/interfaces/app-slice.interface'

export enum APP_SLICE_LOADING {
  BASE = 'isLoading',
  DETAIL = 'isDetailLoading',
  ACTION = 'isActionLoading',
  DOWNLOAD = 'isDownloadLoading'
}

export const APP_SLICE_INITIAL_STATE: IAppSliceInitialState = {
  app_isInitialized: false,
  app_isConnected: true
}
