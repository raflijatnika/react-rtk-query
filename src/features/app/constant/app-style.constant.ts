export enum APP_COLOR {
  DARK = '#06152B',
  WHITE = '#fff'
}

export enum APP_COLOR_LIGHT {
  PRIMARY = '#3A36DB',
  BACKGROUND = '#F1F4FA'
}
