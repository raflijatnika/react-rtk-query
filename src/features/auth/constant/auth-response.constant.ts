// Interfaces
import {
  IAuthResponseLogin,
  IAuthResponseAuthenticatedUser
} from '@/features/auth/interfaces/auth-response.interface'

export const AUTH_RESPONSE_LOGIN: IAuthResponseLogin = {
  token: 'asdf',
  refreshToken: 'asdf'
}

export const AUTH_RESPONSE_AUTHENTICATED_USER: IAuthResponseAuthenticatedUser =
  {
    id: 'asdf',
    name: 'John Doe',
    email: 'johndoe@gmail.com'
  }
