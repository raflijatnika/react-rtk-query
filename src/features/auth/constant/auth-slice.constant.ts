// Interfaces
import { IAuthSliceInitialState } from '@/features/auth/interfaces/auth-slice.interface'

export const AUTH_SLICE_INITIAL_STATE: IAuthSliceInitialState = {
  auth_authenticatedUser: {
    id: null,
    name: null,
    email: null
  },
  auth_token: {
    token: null,
    refreshToken: null
  }
}
