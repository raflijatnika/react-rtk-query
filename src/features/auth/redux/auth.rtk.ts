// Interfaces
import { IAuthAttrsLogin } from '@/features/auth/interfaces/auth-attrs.interface'
import {
  IAuthResponseToken,
  IAuthResponseAuthenticatedUser
} from '@/features/auth/interfaces/auth-response.interface'

// Rtk
import { emptySplitApi } from '@/features/app/redux/app.rtk'

// Mutations
import { auth_SET_TOKEN } from './auth.slice'

export const authApi = emptySplitApi.injectEndpoints({
  endpoints: builder => ({
    auth_register: builder.mutation<
      IAuthResponseAuthenticatedUser,
      IAuthAttrsLogin
    >({
      query: ({ body }) => ({
        url: '/auth/register',
        method: 'POST',
        body
      })
    }),
    auth_login: builder.mutation<IAuthResponseToken['result'], IAuthAttrsLogin>(
      {
        query: ({ body }) => ({
          url: '/auth/login',
          method: 'POST',
          body
        }),
        transformResponse: (response: IAuthResponseToken) => response.result,
        onQueryStarted: async (payload, { dispatch, queryFulfilled }) => {
          try {
            const response = await queryFulfilled

            dispatch(auth_SET_TOKEN(response.data))
          } catch {
            //
          }
        }
      }
    ),
    auth_me: builder.query<IAuthResponseAuthenticatedUser['result'], void>({
      query: () => ({
        url: `/auth/me`
      })
    })
  }),
  overrideExisting: false
})

export const {
  useAuth_registerMutation,
  useAuth_loginMutation,
  useAuth_meQuery
} = authApi
