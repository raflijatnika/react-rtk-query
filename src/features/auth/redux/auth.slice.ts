// Redux Toolkit
import { createSlice, PayloadAction } from '@reduxjs/toolkit'

// Constant
import { AUTH_SLICE_INITIAL_STATE } from '@/features/auth/constant/auth-slice.constant'

// Interfaces
import {
  IAuthToken,
  IAuthAuthenticatedUser
} from '@/features/auth/interfaces/auth.interface'

const auth = createSlice({
  name: 'auth',
  initialState: AUTH_SLICE_INITIAL_STATE,
  reducers: {
    auth_SET_TOKEN: (
      state,
      { payload: { token, refreshToken } }: PayloadAction<IAuthToken>
    ): void => {
      state.auth_token.token = token
      state.auth_token.refreshToken = refreshToken
    },
    auth_SET_AUTHENTICATED_USER: (
      state,
      { payload }: PayloadAction<IAuthAuthenticatedUser>
    ): void => {
      state.auth_authenticatedUser = payload
    },
    auth_LOGOUT: state => {
      state.auth_token = AUTH_SLICE_INITIAL_STATE.auth_token
      state.auth_authenticatedUser =
        AUTH_SLICE_INITIAL_STATE.auth_authenticatedUser
    }
  }
})

// Mutations
export const { auth_SET_TOKEN, auth_SET_AUTHENTICATED_USER, auth_LOGOUT } =
  auth.actions

export default auth.reducer
