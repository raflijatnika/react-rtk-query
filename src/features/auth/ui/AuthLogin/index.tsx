// React
import { useCallback } from 'react'

// Components
import { StyledCredentialsContainer } from '@/features/auth/components'
import {
  AppBaseButton,
  AppBaseDivider,
  AppBaseFormItem,
  AppBaseLabel,
  AppBaseInput,
  AppBaseCheckbox,
  AppBaseSocialMediaButton
} from '@/features/app/components'

// i18n
import { useTranslation } from 'react-i18next'

// Constant
import { APP_COLOR_LIGHT } from '@/features/app/constant/app-style.constant'

// Antd
import { Form } from 'antd'

// React Router DOM
import { useNavigate } from 'react-router-dom'

// Rtk
import { useAuth_loginMutation } from '@/features/auth/redux/auth.rtk'

// Interfaces
import { IAuthLoginForm } from '@/features/auth/interfaces/auth.interface'

const AuthLogin = () => {
  // Hook
  const { t } = useTranslation()
  const [form] = Form.useForm()
  const navigate = useNavigate()

  // Login
  const [auth_login, { isLoading: isLoginLoading }] = useAuth_loginMutation()

  /**
   * @description Redirect to register
   *
   * @return {void} void
   */
  const onRedirectToRegister = useCallback((): void => {
    navigate('/auth/register')
  }, [navigate])

  /**
   * @description Redirect to reset password
   *
   * @return {void} void
   */
  const onRedirectToResetPassword = useCallback((): void => {
    navigate('/auth/reset-password')
  }, [navigate])

  /**
   * @description Login handler
   *
   * @param {IAuthLoginForm} form
   *
   * @return {Promise<void>} Promise<void>
   */
  const onSubmit = useCallback(
    async (form: IAuthLoginForm): Promise<void> => {
      try {
        await auth_login({ body: form }).unwrap()
      } catch (_) {
        //
      }
    },
    [auth_login]
  )

  return (
    <div data-testid='auth-login'>
      {/* Social Media Button */}
      <AppBaseSocialMediaButton />

      {/* Divider */}
      <div className='mt-4 w-full'>
        <AppBaseDivider />
      </div>

      {/* Form */}
      <Form
        form={form}
        onFinish={onSubmit}
        layout='vertical'
        requiredMark={false}
      >
        {/* Email */}
        <AppBaseFormItem
          name='email'
          label={t('app.user.email')}
          rules={[{ required: true, type: 'email' }]}
        >
          <AppBaseInput placeholder='example@gmail.com' />
        </AppBaseFormItem>

        {/* Password */}
        <AppBaseFormItem
          name='password'
          label={t('app.user.password')}
          rules={[{ required: true, type: 'string' }]}
        >
          <AppBaseInput type='password' placeholder={t('app.user.password')} />
        </AppBaseFormItem>

        {/* Remember Me & Reset Password  */}
        <StyledCredentialsContainer>
          {/* Checkbox Of Remember Me */}
          <AppBaseFormItem name='rememberMe' valuePropName='checked'>
            <AppBaseCheckbox>
              <AppBaseLabel fontSize={14.22} fontWeight={400}>
                {t('auth.rememberMe')}?
              </AppBaseLabel>
            </AppBaseCheckbox>
          </AppBaseFormItem>

          {/* Reset Password Link */}
          <AppBaseFormItem>
            <AppBaseLabel
              fontSize={14.22}
              fontWeight={400}
              fontColor={APP_COLOR_LIGHT.PRIMARY}
              className='cursor-pointer'
              onClick={onRedirectToResetPassword}
              data-testid='reset-password'
            >
              {t('auth.resetPassword')}?
            </AppBaseLabel>
          </AppBaseFormItem>
        </StyledCredentialsContainer>

        {/* Login Button */}
        <AppBaseButton
          type='primary'
          htmlType='submit'
          height={50}
          loading={isLoginLoading}
          block
        >
          {t('auth.login')}
        </AppBaseButton>

        {/* Information if user didn't have account */}
        <div className='flex items-center gap-1 mt-5 justify-center'>
          <AppBaseLabel fontSize={16}>{t('auth.notHaveAccount')}?</AppBaseLabel>
          <AppBaseLabel
            fontColor={APP_COLOR_LIGHT.PRIMARY}
            className='cursor-pointer'
            onClick={onRedirectToRegister}
            fontSize={16}
            data-testid='register-button'
          >
            {t('auth.newAccount')}
          </AppBaseLabel>
        </div>
      </Form>
    </div>
  )
}

export { AuthLogin }
