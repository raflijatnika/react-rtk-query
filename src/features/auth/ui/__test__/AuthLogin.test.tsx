// Test
import { act, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import 'jest-styled-components'

// UI
import { AuthLogin } from '../AuthLogin'

// Plugins
import { testRedux_renderWithProviders } from '@/plugins/redux/test.redux'

// Constant
import { AUTH_SLICE_INITIAL_STATE } from '@/features/auth/constant/auth-slice.constant'

// React Router DOM
import { useNavigate } from 'react-router-dom'

it('AuthLogin should render', () => {
  testRedux_renderWithProviders(<AuthLogin />, {
    useRouting: true,
    preloadedState: {
      auth: {
        ...AUTH_SLICE_INITIAL_STATE,
        _persist: { version: 1, rehydrated: true }
      }
    }
  })
})

it('AuthLogin should have snapshot', () => {
  const { container } = testRedux_renderWithProviders(<AuthLogin />, {
    useRouting: true,
    preloadedState: {
      auth: {
        ...AUTH_SLICE_INITIAL_STATE,
        _persist: { version: 1, rehydrated: true }
      }
    }
  })

  expect(container.innerHTML).toMatchSnapshot()
})

it('AuthLogin onRedirectToRegister function should defined', async () => {
  await act(() => {
    testRedux_renderWithProviders(<AuthLogin />, {
      useRouting: true
    })
  })

  const user = userEvent.setup()

  await user.click(screen.getByTestId('register-button'))

  const navigate = useNavigate()

  expect(navigate).toHaveBeenCalled()
})

it('AuthLogin onRedirectToResetPassword function should defined', async () => {
  await act(() => {
    testRedux_renderWithProviders(<AuthLogin />, {
      useRouting: true
    })
  })

  const user = userEvent.setup()

  await user.click(screen.getByTestId('reset-password'))

  const navigate = useNavigate()

  expect(navigate).toHaveBeenCalled()
})

it('AuthLogin onSubmit function should defined', async () => {
  //
})
