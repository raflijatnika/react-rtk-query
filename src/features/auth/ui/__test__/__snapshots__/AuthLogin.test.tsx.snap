// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthLogin should have snapshot 1`] = `"<div data-testid=\\"auth-login\\"><div data-testid=\\"wrapper\\" class=\\"sc-iBkjds kcPEiM\\"><button data-testid=\\"google\\" color=\\"#06152B\\" type=\\"button\\" class=\\"ant-btn ant-btn-default sc-dkzDqf jOMXzK\\"><img src=\\"google.png\\" width=\\"24px\\" height=\\"24px\\"><span>app.social.google</span></button><button data-testid=\\"facebook\\" color=\\"#06152B\\" type=\\"button\\" class=\\"ant-btn ant-btn-default sc-dkzDqf jOMXzK\\"><img src=\\"facebook.png\\" width=\\"9px\\" height=\\"18px\\"><span>app.social.facebook</span></button></div><div class=\\"mt-4 w-full\\"><div class=\\"ant-divider ant-divider-horizontal sc-hKMtZM hBbZGS\\" theme=\\"[object Object]\\" role=\\"separator\\"></div></div><form class=\\"ant-form ant-form-vertical ant-form-hide-required-mark\\"><div class=\\"ant-form-item sc-eCYdqJ cFdIic\\"><div theme=\\"[object Object]\\" class=\\"ant-row ant-form-item-row\\"><div class=\\"ant-col ant-form-item-label\\"><label for=\\"email\\" class=\\"ant-form-item-required\\" title=\\"app.user.email\\">app.user.email</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><input placeholder=\\"example@gmail.com\\" id=\\"email\\" class=\\"ant-input sc-jSMfEi dcJQax\\" type=\\"text\\" value=\\"\\"></div></div></div></div></div><div class=\\"ant-form-item sc-eCYdqJ cFdIic\\"><div theme=\\"[object Object]\\" class=\\"ant-row ant-form-item-row\\"><div class=\\"ant-col ant-form-item-label\\"><label for=\\"password\\" class=\\"ant-form-item-required\\" title=\\"app.user.password\\">app.user.password</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><input type=\\"password\\" placeholder=\\"app.user.password\\" id=\\"password\\" class=\\"ant-input sc-jSMfEi dcJQax\\" value=\\"\\"></div></div></div></div></div><div class=\\"sc-bczRLJ hvLtLW\\"><div class=\\"ant-form-item sc-eCYdqJ cFdIic\\"><div theme=\\"[object Object]\\" class=\\"ant-row ant-form-item-row\\"><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><label class=\\"ant-checkbox-wrapper ant-checkbox-wrapper-in-form-item sc-gKXOVf\\"><span class=\\"ant-checkbox\\"><input id=\\"rememberMe\\" type=\\"checkbox\\" class=\\"ant-checkbox-input\\" value=\\"\\"><span class=\\"ant-checkbox-inner\\"></span></span><span><label font-size=\\"14.22\\" font-weight=\\"400\\" class=\\"sc-gsnTZi cDfIpT\\">auth.rememberMe?</label></span></label></div></div></div></div></div><div class=\\"ant-form-item sc-eCYdqJ cFdIic\\"><div theme=\\"[object Object]\\" class=\\"ant-row ant-form-item-row\\"><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><label font-size=\\"14.22\\" font-weight=\\"400\\" class=\\"sc-gsnTZi gERqLw cursor-pointer\\" data-testid=\\"reset-password\\">auth.resetPassword?</label></div></div></div></div></div></div><button type=\\"submit\\" class=\\"ant-btn ant-btn-primary ant-btn-block sc-dkzDqf leEBRr\\"><span>auth.login</span></button><div class=\\"flex items-center gap-1 mt-5 justify-center\\"><label font-size=\\"16\\" class=\\"sc-gsnTZi llsaSf\\">auth.notHaveAccount?</label><label class=\\"sc-gsnTZi gnknbY cursor-pointer\\" font-size=\\"16\\" data-testid=\\"register-button\\">auth.newAccount</label></div></form></div>"`;
